lib/jobs/ServiceInstanceUpdateJob.js

```diff
@@ -0,0 +1,265 @@
+      .try(() => {
+        if (!skipForbiddenCheck) {
+          this.isChangesPresentInForbiddenSections(diff);
+        }
+      })
+      .then(() => Fabrik
+        .createOperation('update', {
+          deployment: deploymentName,
+          username: 'Auto_Update_Job',
+          arguments: {}
+        })
+        .invoke())
+      .then(result => ({
+        guid: result.guid
+      }));
+  }
+
+  static isChangesPresentInForbiddenSections(diff) {
+    const forbiddenSections = _
+      .chain(diff)
+      .map(_.first)
+      .filter(line => /^[a-z]\w+:/.test(line))
+      .map(line => _.nth(/^([a-z]\w+):/.exec(line), 1))
+      .difference([
+        'update',
+        'compilation',
+        'releases',
+        'resource_pools',
+        'networks',
+        'properties'
+      ])
+      .value();
+    if (!_.isEmpty(forbiddenSections) && !_.includes(forbiddenSections, 'director_uuid')) {
+      throw new errors.Forbidden(`Automatic update not possible. Detected changes in forbidden section(s) '${forbiddenSections.join(',')}'`);
+    }
+    return false;
+  }
+
+  static rescheduleUpdateJob(instanceDetails, trackAttempts) {
+    return Promise.try(() => {
+      const jobData = _.cloneDeep(instanceDetails);
+      jobData.attempt = jobData.attempt + 1;
+      if (trackAttempts) {
+        if (jobData.attempt > config.scheduler.jobs.service_instance_update.max_attempts) {
+          logger.error(`Auto udpate for instance ${instanceDetails.instance_id}  has exceeded max configured attempts : ${config.scheduler.jobs.service_instance_update.max_attempts}}`);
+          return;
+        }
+      }
+      logger.info(`Schedulding InstanceUpdate Job for ${instanceDetails.instance_name}:${instanceDetails.instance_id} @ ${config.scheduler.jobs.reschedule_delay} - Track : ${trackAttempts} - Attempt - ${instanceDetails.attempt}`);
+      return ScheduleManager
+        .runAt(instanceDetails.instance_id,
+          CONST.JOB.SERVICE_INSTANCE_UPDATE,
+          config.scheduler.jobs.reschedule_delay,
+          jobData,
+          CONST.SYSTEM_USER
+        );
+    });
+  }
+
+  static getRandomRepeatInterval() {
+    const afterHr = _.get(config, 'scheduler.jobs.service_instance_update.should_start_after_hr', 0);
+    const beforeHr = _.get(config, 'scheduler.jobs.service_instance_update.should_start_before_hr', 23);
+    const afterMin = _.get(config, 'scheduler.jobs.service_instance_update.should_start_after_min', 0);
+    const beforeMin = _.get(config, 'scheduler.jobs.service_instance_update.should_start_before_min', 59);
+    const runOnceEvery = _.get(config, 'scheduler.jobs.service_instance_update.run_every_xdays', 15);
+    const opts = {
+      start_after_hr: afterHr,
+      start_before_hr: beforeHr,
+      start_after_min: afterMin,
+      start_before_min: beforeMin
+    };
+    return utils.getRandomCronForOnceEveryXDays(runOnceEvery, opts);
+  }
+
+  static getLastRunStatus(name) {
+    const sortOn = ['createdAt', 'asc'];
```

Comment:
shouldnt the sort criteria be desc for getting the last run ?

Reply:
Yes. You are right. Initial logic required asc, but subsequently that was modified but the sort order was never changed. Good catch. Fixed it.